### ISSUE TITLE
bug: File read_to_string, read_to_end does not work on Windows

### ISSUE THREAD

**Issue by robrennie** [WasmEdge/WasmEdge#2888]
### Summary

The Rust "Access OS services" (via WASI) sample doesn't work on Windows. The "read_to_string" function in the sample returns the error "Function not implemented (os error 52)".

### Reproduction steps

```bash
Simply follow the sample instructions for the WASI sample project.
```

### Screenshots

```bash
![DESCRIPTION](LINK.png)
```

### Components

CLI

### WasmEdge Version or Commit you used

0.13.4

### Operating system information

Windows Server 2022

### Hardware Architecture

x64-based processor

**Comment by hydai** [comment]
For those who want to reproduce this issue, the link to the above example: https://wasmedge.org/docs/develop/rust/os/#reading-and-writing-files

**Comment by anshikavashistha** [comment]
@hydai @robrennie We might use conditional compilation to handle file reading differently on Windows and Unix-like systems:

**Comment by anshikavashistha** [comment]
```c
use std::fs::File;
use std::io::{self, Read};

#[cfg(target_os = "windows")]
fn read_file(path: &str) -> io::Result<String> {
    let mut file = File::open(path)?;
    let mut content = String::new();
    file.read_to_string(&mut content)?;
    Ok(content)
}

#[cfg(not(target_os = "windows"))]
fn read_file(path: &str) -> io::Result<String> {
    let mut file = File::open(path)?;
    let mut content = Vec::new();
    file.read_to_end(&mut content)?;
    Ok(String::from_utf8_lossy(&content).to_string())
}

fn main() {
    let path = "your_file_path_here";
    match read_file(path) {
        Ok(content) => println!("File content: {}", content),
        Err(e) => eprintln!("Error reading file: {}", e),
    }
}

```

**Comment by dannypsnl** [comment]
https://doc.rust-lang.org/std/fs/fn.read_to_string.html#

**Comment by anshikavashistha** [comment]
Alright @dannypsnl .Then what kind of approach should we consider?

**Comment by TBBle** [comment]
tl;dr: Fixed in 0.14.1 by #3474 "Treat EOF error as recoverable". 0.14.1 is not yet available via [winget](https://wasmedge.org/docs/start/install/#windows), but pulling the MSI file directly from [the release](https://github.com/WasmEdge/WasmEdge/releases/tag/0.14.1) and installing it works.

I've also raised [an issue to have the package upgraded in winget](https://github.com/microsoft/winget-pkgs/issues/175820).

<details>

----

<summary>Record of investigation with 0.14.0, the latest version available on winget</summary>

I can still reproduce the issue with:
* rustc 1.81.0 (eeb90cda1 2024-09-04); and either
* C:\Program Files\WasmEdge\bin\wasmedge.exe version 0.13.5; or
* C:\Program Files\WasmEdge\bin\wasmedge.exe version 0.14.0

<details>

<summary>Code from the example, but unwrapping the failing line and a couple of clippy lints fixed</summary>

----

(Note that by using unwrap this way, I get a clearer failure but the actual string returned would be the number of bytes read, not the contents.)

```rust
// https://wasmedge.org/docs/develop/rust/os/#reading-and-writing-files
// https://github.com/WasmEdge/WasmEdge/issues/2888
use rand::prelude::*;

pub fn get_random_i32() -> i32 {
    let x: i32 = random();
    x
}

pub fn get_random_bytes() -> Vec<u8> {
    let mut rng = thread_rng();
    let mut arr = [0u8; 128];
    rng.fill(&mut arr[..]);
    arr.to_vec()
}

pub fn echo(content: &str) -> String {
    println!("Printed from wasi: {}", content);
    content.to_string()
}

use std::env;

pub fn print_env() {
    println!("The env vars are as follows.");
    for (key, value) in env::vars() {
        println!("{}: {}", key, value);
    }

    println!("The args are as follows.");
    for argument in env::args() {
        println!("{}", argument);
    }
}

use std::fs;
use std::fs::File;
use std::io::{Read, Write};

pub fn create_file(path: &str, content: &str) {
    let mut output = File::create(path).unwrap();
    output.write_all(content.as_bytes()).unwrap();
}

pub fn read_file(path: &str) -> String {
    let mut f = File::open(path).unwrap();
    let mut s = String::new();
    f.read_to_string(&mut s).unwrap().to_string()
}

pub fn del_file(path: &str) {
    fs::remove_file(path).expect("Unable to delete");
}

fn main() {
    println!("Random number: {}", get_random_i32());
    println!("Random bytes: {:?}", get_random_bytes());
    println!("{}", echo("This is from a main function"));
    print_env();
    create_file("tmp.txt", "This is in a file");
    println!("File content is {}", read_file("tmp.txt"));
    del_file("tmp.txt");
}
```

----

</details>

```console
> $Env:CARGO_TARGET_WASM32_WASIP1_RUNNER="wasmedge --dir .:. --env RUST_BACKTRACE=full"
> cargo run --target wasm32-wasip1
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.09s
     Running `wasmedge --dir .:. --env RUST_BACKTRACE=full target\wasm32-wasip1\debug\wasmedge-2888.wasm`
Random number: -2109017026
Random bytes: [79, 210, 103, 88, 43, 182, 0, 74, 1, 121, 235, 191, 159, 125, 253, 115, 221, 41, 173, 3, 241, 192, 149, 44, 238, 191, 137, 239, 221, 119, 140, 179, 123, 240, 5, 41, 183, 1, 137, 77, 233, 229, 209, 109, 82, 116, 231, 118, 31, 225, 123, 204, 192, 85, 30, 113, 193, 120, 127, 77, 75, 114, 161, 121, 22, 48, 45, 236, 62, 195, 213, 20, 36, 252, 25, 69, 249, 159, 123, 98, 53, 250, 254, 36, 216, 158, 162, 150, 37, 6, 138, 36, 96, 199, 233, 188, 220, 198, 101, 121, 171, 45, 38, 253, 114, 133, 57, 236, 53, 94, 148, 164, 0, 61, 109, 53, 244, 245, 26, 245, 139, 91, 211, 5, 182, 181, 28, 27]
Printed from wasi: This is from a main function
This is from a main function
The env vars are as follows.
RUST_BACKTRACE: full
The args are as follows.
wasmedge-2888.wasm
thread 'main' panicked at src/main.rs:48:30:
called `Result::unwrap()` on an `Err` value: Os { code: 52, kind: Unsupported, message: "Function not implemented" }
stack backtrace:
[2024-09-30 13:28:40.785] [error] execution failed: unreachable, Code: 0x89
[2024-09-30 13:28:40.787] [error]     In instruction: unreachable (0x00) , Bytecode offset: 0x000414a5
[2024-09-30 13:28:40.787] [error]     When executing function name: "_start"
```

I can confirm the `tmp.txt` file is successfully being created, and if I just use `cargo run` it works.

If I change `read_file` to 

```rust
pub fn read_file(path: &str) -> String {
    let mut f = File::open(path).unwrap();
    assert_eq!(17, f.metadata().unwrap().len());
    let mut buf = [0; 17];
    f.read_exact(&mut buf).unwrap();
    String::from_utf8(buf.to_vec()).unwrap()
}
```

it works, but 

```rust
pub fn read_file(path: &str) -> String {
    let mut f = File::open(path).unwrap();
    assert_eq!(17, f.metadata().unwrap().len());
    let mut buf = [0; 18];
    f.read_exact(&mut buf).unwrap();
    String::from_utf8(buf.to_vec()).unwrap()
}
```
fails with the same error.

```rust
pub fn read_file(path: &str) -> String {
    let mut f = File::open(path).unwrap();
    assert_eq!(17, f.metadata().unwrap().len());
    let mut buf = [0; 8192];
    let read_bytes = f.read(&mut buf).unwrap();
    assert_eq!(read_bytes, f.metadata().unwrap().len() as usize);
    assert_eq!(0, f.read(&mut buf).unwrap());
    String::from_utf8(buf.to_vec()).unwrap()
}
```

fails on the _second_ read, that should return 0 (which is how `read_exact` is implemented), so I think the issue is a failure when attempting to read from an EOF file, in [`INode::fdRead`](https://github.com/WasmEdge/WasmEdge/blob/0.14.0/lib/host/wasi/inode-win.cpp#L1047-L1095).

When reading past EOF, [`ReadFileEx`](https://learn.microsoft.com/en-au/windows/win32/api/fileapi/nf-fileapi-readfileex#return-value) returns `ERROR_HANDLE_EOF` for the [`GetOverlappedResult`'s `GetLastError`](https://github.com/WasmEdge/WasmEdge/blob/0.14.0/lib/host/wasi/inode-win.cpp#L1081) and there's no handling for it there, so [it propagates up as `__WASI_ERRNO_NOSYS`](https://github.com/WasmEdge/WasmEdge/blob/0.14.0/lib/host/wasi/win.h#L196-L202).

As it happens, [the 0.14.1 version of `INode::fdRead` handles this error case](https://github.com/WasmEdge/WasmEdge/blob/0.14.1/lib/host/wasi/inode-win.cpp#L1078-L1080)... So it's actually been fixed already, but `winget upgrade wasmedge` didn't pick it up. `>_<`

</details>

**Comment by hydai** [comment]
Thanks for the confirmation from @TBBle 
Closing this.
